tests/common/include/tests/common/tools_stepper.inl

```diff
@@ -217,3 +219,97 @@ TEST(ALGEBRA_PLUGIN, rk_stepper) {
         }
     }
 }
+
+// This tests the covariance transport in rk stepper
+TEST(ALGEBRA_PLUGIN, covariance_transport) {
+
+    using vector3 = vector3<scalar>;
+    using point3 = point3<scalar>;
+    using matrix_operator = standard_matrix_operator<scalar>;
+    using mag_field_t = constant_magnetic_field<>;
+    using crk_stepper_t =
+        rk_stepper<mag_field_t, free_track_parameters, constrained_step<>>;
+
+    // Generate track starting point
+    point3 pos{0., 0., 0.};
+    scalar time = 0.;
+    vector3 mom{0.01, 0.01, 0.01};
+    scalar q = -1.;
+
+    free_track_parameters vertex(pos, time, mom, q);
+
+    // Fill covariance matrix
+    auto covariance0 =
+        matrix_operator().template zero<e_free_size, e_free_size>();
+
+    matrix_operator().element(covariance0, e_free_pos0, e_free_pos0) = 0.1;
+    matrix_operator().element(covariance0, e_free_pos1, e_free_pos1) = 0.1;
+    matrix_operator().element(covariance0, e_free_pos2, e_free_pos2) = 0.1;
+    matrix_operator().element(covariance0, e_free_dir0, e_free_dir0) = 0.1;
+    matrix_operator().element(covariance0, e_free_dir1, e_free_dir1) = 0.1;
+    matrix_operator().element(covariance0, e_free_dir2, e_free_dir2) = 0.;
+    matrix_operator().element(covariance0, e_free_qoverp, e_free_qoverp) = 0.;
+
+    vertex.set_covariance(covariance0);
+
+    // magnetic field
+    vector3 B{0, 0, 10. * unit_constants::T};
+
+    // helix gun
+    helix_gun helix(vertex, B);
+
+    // Path length per turn
+    scalar S = 2. * getter::norm(mom) / getter::norm(B) * M_PI;
+
+    // RK stepper and its state
+    mag_field_t mag_field(B);
+    crk_stepper_t::state crk_state(vertex);
+
+    // Decrease tolerance down to 1e-8
+    crk_state._tolerance = 1e-8;
+
+    crk_stepper_t crk_stepper(mag_field);
+
+    // Run stepper for one turn
+    unsigned int max_steps = 1e4;
+    for (unsigned int i = 0; i < max_steps; i++) {
+
+        crk_state.set_constraint(S - crk_state.path_length());
+
+        n_state._step_size = std::numeric_limits<scalar>::max();
+
+        crk_stepper.step(crk_state, n_state);
+
+        if (std::abs(S - crk_state.path_length()) < 1e-6) {
+            break;
+        }
+
+        // Make sure that we didn't reach the end of for loop
+        ASSERT_TRUE(i < max_steps);
```

Comment:
Isn't this checked in the loop, or am I missing something? Once we break the loop we will never arrive here anyways

Reply:
Right. I will change `i < max_steps-1`